core/src/main/kotlin/org/evomaster/core/problem/api/service/ApiWsStructureMutator.kt

```diff
@@ -25,56 +27,126 @@ import kotlin.math.min
 /**
  * the abstract structure mutator for API based SUT, such as REST, GraphQL, RPC
  */
-abstract class ApiWsStructureMutator : StructureMutator(){
+abstract class ApiWsStructureMutator : StructureMutator() {
 
     companion object {
         private val log: Logger = LoggerFactory.getLogger(ApiWsStructureMutator::class.java)
     }
 
+    // TODO: This will moved under ApiWsFitness once RPC and GraphQL support is completed
+    @Inject
+    protected lateinit var externalServiceHandler: ExternalServiceHandler
 
-    @Deprecated("External Actions will be moved into EnterpriseActionGroup")
-    private fun <T : ApiWsIndividual> addInitializingExternalServiceActions(
+    private fun addExternalServiceActions(
         individual: EvaluatedIndividual<*>,
+        /**
+         * TODO add why
+         */
         mutatedGenes: MutatedGeneSpecification?,
-        sampler: ApiWsSampler<T>
     ) {
-        // TODO: Incomplete code, under development
 
         if (config.externalServiceIPSelectionStrategy == EMConfig.ExternalServiceIPSelectionStrategy.NONE) {
             return
         }
 
-        if (sampler is ResourceSampler) {
-            val ind = individual.individual as? T
-                ?: throw IllegalArgumentException("Invalid individual type")
+        val ind = individual.individual as? ApiWsIndividual
+            ?: throw IllegalArgumentException("Invalid individual type")
+
+        val esr = individual.fitness.getViewAccessedExternalServiceRequests()
+        if (esr.isEmpty()) {
+            //nothing to do
+            return
+        }
 
-            // TODO: Under review
-            val fw = individual.fitness.getAccessedExternalServiceRequests()
+        val newActions: MutableList<HttpExternalServiceAction> = mutableListOf()
 
-            val actions = mutableListOf<ExternalServiceAction>().plus(
-                sampler.getExternalService().getExternalServiceActions()
-            )
+        ind.seeMainExecutableActions().forEachIndexed { index, action ->
+            val parent = action.parent
+            if (parent !is EnterpriseActionGroup) {
+                //TODO this should not really happen
+                val msg = "Action is not inside an EnterpriseActionGroup"
+                log.error(msg)
+                throw RuntimeException(msg)
+            }
 
-            ind.addInitializingDbActions(0, actions)
+            if (esr.containsKey(index)) {
```

Comment:
as we discussed, add a doc about here we only handling of adding new external services here. regarding the external service actions which are not used, we handled them after fitness evaluation with `used` property